`Development`: Adds basis for detecting flaky e2e tests with coverage

heddendorp opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).


I have not changed actual application code.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
As part of my thesis I want to bring the flaky test detecting version of E2E tests online in the regular Artemis development flow. Currently it should be run in parallel to see if there might be unintended disadvantages such as more failures or longer build times.

### Description
<!-- Describe your changes in detail -->
Adds a new docker compose and related files to run E2E tests with full coverage across both server and client. This allows detecting flaky tests by comparing the covered files to changed files since the last time the build was green.

All changes where made in a way that ensures, the regular E2E tests are not affected [check them here](https://bamboobruegge.in.tum.de/browse/ARTEMIS-AETG1114)

You can check the flake detection enabled plan [here](https://bamboobruegge.in.tum.de/browse/ARTEMIS-AECF4)

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![image](https://user-images.githubusercontent.com/3612748/206911784-ed990dc6-eec3-4eb8-b980-c13be6b54e2b.png)
Changes allow tagging failing Bamboo builds as suspected flaky and will report a positive status to GitHub if all fails are suspected to be flaky.
![image](https://user-images.githubusercontent.com/3612748/206911920-02336819-2df1-4a9b-95bf-9d02851b9bbb.png)
This PR already includes the flake detection plan and will report it's feedback

krusche commented:
Adding jars to the repo is not recommended. Could we also add this as test dependency?

heddendorp commented:
> Adding jars to the repo is not recommended. Could we also add this as test dependency?

That's actually something I wanted to look into after the general idea works, they are needed during regular execution so a test dependency is probably the wrong place.
I'm not too familiar with how I can add them as a normal dependency but will make sure to have a look before moving this out of draft.

heddendorp commented:
@krusche I have removed the jars from the repo and added the jacoco agent as a regular dependency.
For the flaky tests I extract it from the Artemis .war file and attach it as a java agent.

Other changes in this PR should be good to go. The regular E2E tests are still passing and are not affected

4ludwig4 commented on bootstrap-coverage.sh:
```diff
@@ -0,0 +1,37 @@
+  --add-opens java.base/sun.nio.ch=ALL-UNNAMED \
+  --add-opens java.management/sun.management=ALL-UNNAMED \
+  --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED \
+  -javaagent:jacocoagent.jar=output=tcpserver,address=* \
```
If we really want to put this into the docker compose file you would have to create a very long chain of commands which is do-able, but I prefer the readability of an extra script as long as we need multiple steps and also steps before the java command as you do. So this extra file is okay.

4ludwig4 commented on src/main/docker/Dockerfile:
```diff
@@ -8,7 +10,7 @@ ENV SPRING_OUTPUT_ANSI_ENABLED=ALWAYS \
 
 RUN echo "Installing needed dependencies" \
-  && apt-get update && apt-get install -y --no-install-recommends locales graphviz wget \
+  && apt-get update && apt-get install -y --no-install-recommends locales graphviz wget unzip \
```
Mhh not sure if we can get rid of the unzip as it's not necessary for production systems of Artemis? At least it has to be documented what this is used for.

heddendorp replied:
@4ludwig4 I have removed any changes to the Dockerfile

4ludwig4 commented on src/main/docker/Dockerfile:
```diff
@@ -38,7 +40,7 @@ WORKDIR /opt/artemis
 COPY $WAR_PATH/*.war Artemis.war
 
-COPY bootstrap.sh /bootstrap.sh
+COPY ${BOOTSTRAP_SCRIPT} /bootstrap.sh
```
Out of my gut feeling I would rather mount and overwrite the bootstrap.sh inside the container with your version of it than inserting it via the ARG.

heddendorp replied:
Also changed into separate image

4ludwig4 reviewed (dismissed):
I like the approach with the override Docker Compose file and the Dockerfile extending the Artemis Dockerfile. So from a container point of view this seems right.

You can think of also integrating the bamboo scripts in version control here like I did in:

- #5925

This makes reviewing and debugging easier for other devs.

4ludwig4 commented:
@heddendorp I just had a thought which could lead to problems. So far, I think we are just starting up Artemis instances for the E2E tests on the bamboo agents. If we start up the same named containers also for this new bamboo task, we could potentially run into problems. Not sure how docker names the containers and volumes for the old cypress setup. We should maybe take a look at that before merging.

krusche commented on src/test/cypress/package.json:
```diff
@@ -8,6 +8,7 @@
     ],
     "devDependencies": {
         "@4tw/cypress-drag-drop": "2.1.0",
+        "@heddendorp/coverage-git-compare": "^1.3.1",
```
Please use specific versions and avoid ^

krusche commented on bootstrap-coverage.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
Can we move this file into a suitable sub folder? We already have a lot of files in the root folder which is already cluttering. We should not increase it if we can avoid it

krusche reviewed (changes requested):
See my inline feedback

heddendorp commented:
> @heddendorp I just had a thought which could lead to problems. So far, I think we are just starting up Artemis instances for the E2E tests on the bamboo agents. If we start up the same named containers also for this new bamboo task, we could potentially run into problems. Not sure how docker names the containers and volumes for the old cypress setup. We should maybe take a look at that before merging.

@4ludwig4 I think the bamboo plan runs are sufficiently separated as we run both plans multiple times in parallel already I do not expect any issues.

4ludwig4 commented:
> > @heddendorp I just had a thought which could lead to problems. So far, I think we are just starting up Artemis instances for the E2E tests on the bamboo agents. If we start up the same named containers also for this new bamboo task, we could potentially run into problems. Not sure how docker names the containers and volumes for the old cypress setup. We should maybe take a look at that before merging.
> 
> @4ludwig4 I think the bamboo plan runs are sufficiently separated as we run both plans multiple times in parallel already I do not expect any issues.

Just looked it up and Bamboo seems to just run one task at a time per agent, so we should be good!

Strohgelaender commented on src/main/docker/cypress/Coverage-Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM artemis:coverage-latest
+
+RUN echo "Installing needed dependencies" \
+  && apt-get update && apt-get install -y --no-install-recommends unzip \
```
Maybe this is a misunderstanding on my side (since I saw this missing in multiple docker files in context of Artemis), but from my understanding `update` only checks if new versions are available without actually installing them. To do the actual update you need to also call `upgrade`.
```suggestion
  && apt-get update && apt-get upgrade -y && apt-get install -y --no-install-recommends unzip \
```

heddendorp replied:
This is lifted from other artemis dockerfiles so that might be the reason.
https://github.com/ls1intum/Artemis/blob/b6f6ae519f6f039bd5791a7efccd0382827e70c2/src/main/docker/Dockerfile#L10-L13
Update indeed loads the latest package information, mainly to make sure the package about to be installed is the latest one in this case.
Since I don't actually want to change anything but get this utility, I think it is fine to only update @Strohgelaender

Mtze commented on docs/dev/cypress.rst:
```diff
@@ -107,6 +107,35 @@ of the Cypress test suite, since the new Artemis executable first has to be buil
+----------------------------------------------------
+In addition to our regular Cypress execution, we also run a special experimental build plan that attempts to detect
+flaky tests based on the changed code. To do this, we have some special Docker configurations that are specific to this
+`build plan <https://bamboobruegge.in.tum.de/browse/ARTEMIS-AECF>`__.
```
```suggestion
`build plan <https://bamboo.ase.in.tum.de/browse/ARTEMIS-AECF>`__.
```

Mtze commented on src/main/docker/cypress/Coverage-Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM artemis:coverage-latest
```
I'm not sure where this name and tag comes from. I think it should be 
```suggestion
FROM ghcr.io/ls1intum/artemis:latest
``` 

If your intention was to build and tag the image locally you should put that in the documentation :)

heddendorp replied:
Yes, the setup uses a locally built image. I will update the documentation 👍

heddendorp commented:
@Mtze thanks for your feedback, should be resolved now

TheZoker reviewed (approved):
Re-approving 👍
